xds/src/main/java/io/grpc/xds/internal/sds/ServerWrapperForXds.java

```diff
@@ -77,27 +132,112 @@ public Server start() throws IOException {
       throw new RuntimeException(ex);
     }
     delegate.start();
+    currentServingState = ServingState.STARTED;
     xdsServingStatusListener.onServing();
     return this;
   }
 
-  private Future<EnvoyServerProtoData.DownstreamTlsContext> addServerWatcher() {
-    final SettableFuture<EnvoyServerProtoData.DownstreamTlsContext> settableFuture =
-        SettableFuture.create();
+  @VisibleForTesting public ServingState getCurrentServingState() {
+    return currentServingState;
+  }
+
+  private SettableFuture<Throwable> addServerWatcher() {
+    final SettableFuture<Throwable> future = SettableFuture.create();
     serverWatcher =
         new XdsClientWrapperForServerSds.ServerWatcher() {
           @Override
-          public void onError(Throwable throwable) {
-            xdsServingStatusListener.onNotServing(throwable);
+          public void onError(Throwable throwable, boolean isAbsent) {
+            if (currentServingState == ServingState.STARTING) {
+              // during start
+              if (isPermanentErrorFromXds(throwable)) {
+                currentServingState = ServingState.SHUTDOWN;
+                future.set(throwable);
+                return;
+              }
+              xdsServingStatusListener.onNotServing(throwable);
+            } else {
+              // after start
+              if (isAbsent) {
+                xdsServingStatusListener.onNotServing(throwable);
+                if (currentServingState == ServingState.STARTED) {
```

Comment:
There could be a race between checking the value and `currentServingState` getting updated from `STARTING` to `STARTED`.

Reply:
synchronizing on `ServerWrapperForXds.this` now